### ISSUE TITLE
switch `nil` and `iota` away from booleans

### ISSUE THREAD

**Issue by i1v** [fatih/vim-go#3517]
(sorry for new issue, but previous issue was quickly closed, signalling that further discussion is not welcome)

currently `nil` and `iota` are defined as `goPredefinedIdentifiers`:

https://github.com/fatih/vim-go/blob/3d16fa33a9a1b707a15da3ca4e3f2a7374d0902a/syntax/go.vim#L53

which is then linked to `goBoolean`:

https://github.com/fatih/vim-go/blob/3d16fa33a9a1b707a15da3ca4e3f2a7374d0902a/syntax/go.vim#L57

and then highlighted as `Boolean`:

https://github.com/fatih/vim-go/blob/3d16fa33a9a1b707a15da3ca4e3f2a7374d0902a/syntax/go.vim#L56

in some other code, I have seen it declared like this:

~~~vim
syn keyword     goConstants         iota true false nil
hi def link     goConstants         Keyword
~~~

I think this is a sensible option, as `nil` and `iota` are not booleans.

**Comment by bhcleek** [comment]
> (sorry for new issue, but previous issue was quickly closed, signalling that further discussion is not welcome)

You asked a question and an answer was provided. The syntax change you asked for is not going to be implemented.

> in some other code, I have seen it declared like this:
```
syn keyword     goConstants         iota true false nil
hi def link     goConstants         Keyword
```
> I think this is a sensible option, as nil and iota are not booleans.

None of `iota`, `nil`, `false`, or `true` are keywords either. You're free to use the `Keyword` group for `goBoolean` and/or `goPredefinedIdentifiers` if you'd like.

I understand that you think it's sensible. What vim-go currently does is also sensible. If you'd like to use a different highlighting group for `iota` and `nil`, you're free to  do that in your vim config. They're put into separate highlighting groups so that we can avoid this kind of bike shedding. See https://github.com/fatih/vim-go/issues/3482#issuecomment-1379896262 for some tips on how use a different group's colors than what vim-go does out of the box.

**Comment by i1v** [comment]
> None of `iota`, `nil`, `false`, or `true` are keywords either.

Fair enough, but this comment is actually a deflection, rather than actually addressing the concern. even if my suggestion of keyword is poor, an identifier is not a boolean, so I think the current code could be improved, with something other than boolean or keyword. Do you concede that?

> You're free to use the `Keyword` group for `goBoolean` and/or `goPredefinedIdentifiers` if you'd like.

another deflection, this issue is not about how I can work around an issue with the code, but how I can address the issue directly.

**Comment by bhcleek** [comment]
There are no deflections here. I simply disagree with you. As for your concern: it's the very definition of bike shedding. Yes, an identifer is not a boolean, but the `true`, `false`, `iota`, and `nil` are all identifiers, two of which are used for boolean values. Vim-go long ago (see #1030 and #1049) chose to separate those four into two highlight groups so they could be highlighted differently if a user wants. One could argue about whether `Keyword`, `Boolean`, `Identifier`, or some other group is the right group to link against by default for either of the two vim-go groups, but I'm not terribly interested in that topic, and I'm definitely not interested in spending a lot of time second guessing decisions that were made years ago that you have an easy way to change on your system. (It's worth pointing out that the default coloring for `nil` and `iota` was made to be `Identifier` in #1030 and folks didn't like it, so back to `Boolean` by default it was!)

> this issue is not about how I can work around an issue with the code, but how I can address the issue directly.

This conversation is the very definition of bike shedding. You don't like the color of a couple of words. Modify your configuration to your tastes; vim-go makes it easy for you to do that.

**Comment by i1v** [comment]
I am not asking to change from red to orange, or orange to yellow. I am just asking that we not call `nil` a boolean. I think thats a reasonable request, and not a "bike shed". contrary to your gaslighting, `nil` is not a boolean, never has been, never will be. thats just a simple fact. keeping the current code as is might be a minor issue, but it seems simple enough to fix, and I would even be willing to submit a PR, if I knew that maintain would be open to it

**Comment by bhcleek** [comment]
No one has claimed that `nil` is a boolean. It merely uses the same coloring.

**Comment by i1v** [comment]
sorry I was not clear in my previous comment. the main issue is here:

~~~vim
hi def link     goPredefinedIdentifiers    goBoolean 
~~~

identifiers are not booleans, so I think it would be an improvement if the code didn't represent them as such. I think my previous comments made that clear, but if you wish to continue semantic arguments, I can try to continue improving my concern until it is no longer ambiguous.

**Comment by bhcleek** [comment]
It doesn't represent them as such, though. All this code is saying is, "by default, link the highlighting of the `goPredefinedIdentifiers` group to the highlighting of the `goBoolean` group," so that the highlighting `goBoolean` group and the `goPredefinedIdentifiers` group are the same.

**Comment by bhcleek** [comment]
> I think it would be an improvement if the code didn't represent them as such.

I think I misinterpreted what you meant by this. You're not saying the that vim-go code represents them as such, you're saying that vim-go highlights them to be the same? Is that right? If so, this was already discussed in the #1030 and made the way it is now in #1049 in response to them being highlighted differently in #1030. If you'd like other highlighting, then please leverage `hi default link goPredefinedIdentifiers` or `hi link goPredefinedIdentifiers` and append the group whose highlighting you want the `goPredefinedIdentifiers to use`. e.g. `hi link goPredefinedIdentifiers Identifier`

**Comment by i1v** [comment]
I am going to try starting from scratch here, and you seen to be continually misunderstanding my position, either intentionally or otherwise. So I will give the benefit of the doubt, and try to do a better job of explaining my view.

currently `goPredefinedIdentifiers` are linked to `goBoolean`, which is in turn linked to `Boolean`:

https://github.com/fatih/vim-go/blob/3d16fa33a9a1b707a15da3ca4e3f2a7374d0902a/syntax/go.vim#L53-L57

I think this is a mistake. Identifiers are not booleans. Even in the context of highlighting, I think its a mistake as well. Vim already has two builtins for Identifiers, one for dark and light:

https://github.com/vim/vim/blob/fdbd14e89207dca95549f298ee468793da852086/runtime/syntax/syncolor.vim#L37

https://github.com/vim/vim/blob/fdbd14e89207dca95549f298ee468793da852086/runtime/syntax/syncolor.vim#L48

so perhaps improved code would look like this:

~~~vim
hi def link goPredefinedIdentifiers Identifier
~~~

**Comment by bhcleek** [comment]
I understand, @i1v. My answer remains the same, though. This was already discussed by maintainers and contributors in #1030 and made the way it is now in #1049 after #1030 made it the way you're asking for.

You may not like that the `Boolean` group is used for both `goBoolean` and `goPredefinedIdentifiers`, so vim-go and vim make it easy to customize how `goPredefinedIdentifiers` are highlighted.
